### DISCUSSION TITLE
Rewrite all UIDs

### DISCUSSION THREAD

**Discussion by marsianer** [ical4j/ical4j#646] [Q&A] [answered]
For historical reasons, my calendar has UIDs that do not correspond to the current UUID standard. 
Can I automatically replace all UIDs and then re-import the calendar?

https://de.wikipedia.org/wiki/Universally_Unique_Identifier

Example:
c6ec8f24-47e9-447e-b408-97463af921fe.ics [OK]
libkcal-939447340.715.ics [REPLACE]

**Comment by benfortuna** [comment]
It will probably depend on your calendaring application, but I expect if you import the same events with different UIDs then it will create duplicates of all your events.

So you can possibly try using the METHOD property to indicate the old events should be removed, but I can't think of a way to change the UID of an event in-place.

**Comment by marsianer** [comment]
I would like to set up the calendar from scratch. 
All clients therefore receive a new initial sync.

**Reply by benfortuna** [reply, marked as answer]
I think that should be ok. So to use iCal4j for this you will just read in the existing calendar(s) and iterate through the events and replace existing UID properties with a new one. Then write out the calendar(s) to a file for import.

- Use [Calendar.getComponents()](http://ical4j.github.io/docs/ical4j/api/4.0.0-beta8/net/fortuna/ical4j/model/Calendar.html#getComponents()) to iterate all events
- Use [event.replace(..new UID..)](http://ical4j.github.io/docs/ical4j/api/4.0.0-beta8/net/fortuna/ical4j/model/PropertyContainer.html#replace(net.fortuna.ical4j.model.Property)) to update the event in place
- Use [RandomUidGenerator.generateUid()](http://ical4j.github.io/docs/ical4j/api/4.0.0-beta8/net/fortuna/ical4j/util/RandomUidGenerator.html#generateUid()) to create new UIDs
